[DDP] fix test_ddp_inference

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61666**

Closes https://github.com/pytorch/pytorch/issues/61481. Fixes this
test by removing section that uses only torch.no_grad() and doesn't call
model.eval(). For SyncBN, need to call model.eval() otherwise SyncBN will
assume it is in training mode, which does collective calls in the forward pass
and does not work for inference.

Differential Revision: [D29699444](https://our.internmc.facebook.com/intern/diff/D29699444/)

rohan-varma committed:
[DDP] fix test_ddp_inference

Closes https://github.com/pytorch/pytorch/issues/61481. Fixes this
test by removing section that uses only torch.no_grad() and doesn't call
model.eval(). For SyncBN, need to call model.eval() otherwise SyncBN will
assume it is in training mode, which does collective calls in the forward pass
and does not work for inference.

Differential Revision: [D29699444](https://our.internmc.facebook.com/intern/diff/D29699444/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -7112,14 +7111,8 @@ def test_ddp_inference(self):
             for test in tests:
                 test_model, test_local_model, test_inp = test
                 if self.rank == 0:
-                    with torch.no_grad():
```
Does this mean `no_grad` is no longer relevant for inference and `.eval()` will sufficient for all cases?

rohan-varma committed:
Update on "[DDP] fix test_ddp_inference"


Closes https://github.com/pytorch/pytorch/issues/61481. Fixes this
test by removing section that uses only torch.no_grad() and doesn't call
model.eval(). For SyncBN, need to call model.eval() otherwise SyncBN will
assume it is in training mode, which does collective calls in the forward pass
and does not work for inference.

Differential Revision: [D29699444](https://our.internmc.facebook.com/intern/diff/D29699444/)

[ghstack-poisoned]

rohan-varma commented:
Failure is unrelated - `FAIL [21.893s]: test_fd_sharing (__main__.TestMultiprocessing)`
